Volume preparation using Salt formulas

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'storage', 'kubernetes'

**Context**: 

see:  #1474

**Summary**:

A combination of Salt state and custom salt modules should be able to take 0 or 1 device as arguments and prepare the specified volume if no argument is passed, we should prepare all volumes not yet prepared within the pillars.

**Acceptance criteria**: 

- Be able to successfully prepare loop devices and block devices as metalk8s volumes.
- Ensure that prepared volumes are always available even after a restart.
- Ensure that prepared volumes have a valid file system
- Be able to write persistent data on all prepared volumes 

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1474

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

Ebaneck commented:
@bert-e wait

jbertran commented on salt/metalk8s/volume/init.sls:
```diff
@@ -0,0 +1,6 @@
+#
```
Here a call on `metalk8s.volume` would prepare then unprepare a volume.

Let's leave this as a prepared volume if you don't specify the underlying state.

jbertran commented on salt/metalk8s/volume/files/loopdevice.service.j2:
```diff
@@ -0,0 +1,17 @@
+{# source salt linux formulars #}
```
Typo: `formulas`

jbertran commented on salt/_states/metalk8s_volume.py:
```diff
@@ -0,0 +1,26 @@
+    name
+        the name of the storageclass
+    '''
+    # add valid state definition for this
```
Is that a TODO?

jbertran commented on salt/metalk8s/volume/prepare/init.sls:
```diff
@@ -0,0 +1,84 @@
+  metalk8s_volume.get_storageclass_validity:
+  - name: {{ storageclass_name }}
+
+    {%- if 'sparseLoopDevice' in v_info['spec'].keys() %}
```
More indentation please! It's hard to follow what belongs to the ``sparseLoopDevice` check and what doesn't below.

jbertran commented on salt/metalk8s/volume/prepare/init.sls:
```diff
@@ -0,0 +1,84 @@
+
+Create sparseLoopDevice {{ v_name }} directory:
+  file.directory:
+  - name: /dev/{{ v_name }}
```
Are we OK with crashing in a `file.directory` state if the volume name is already taken? Even if volume name collision checks are handled at a higher level, we should probably check here as well (the usual TOCTOU issues)

jbertran commented on salt/metalk8s/volume/prepare/init.sls:
```diff
@@ -0,0 +1,84 @@
+
+Create sparsefile {{ v_name }}:
+  cmd.run:
+  - name: "truncate --size {{ readable_v_size }} {{ v_name }}"
```
Shouldn't these be `/dev/{{v_name}}/{{v_name}}`, according to your init scripts?

NicolasT replied:
We can't have those sparse files somewhere under `/dev` to begin with... `/dev` is not exactly persistent.

Ebaneck replied:
well noted

NicolasT replied:
Also, don't use `cmd.run` with `truncate --size`. https://github.com/scality/metalk8s/issues/1474#issuecomment-518308958

jbertran commented on salt/_modules/metalk8s_volume.py:
```diff
@@ -0,0 +1,54 @@
+
+# simple helper to convert kubernetes size object to human readable form
+# e.g 43Gi from Kubernetes yields 46170898432
+def to_bytes_converter(quantity):
```
I feel like this is a common enough problem to have an existing importable utility, but my google-fu fails me...

@slaperche-scality ?

NicolasT replied:
Can't find anything in `python-kubernetes`.

slaperche-scality replied:
Not in the stdlib I'm affraid.

As to bring an external dependency for a 10 lines (counting code only) function, I'm not convinced by the JS way of life…

That being said, would be nice to have some unit tests for small helpers like that.
Probably out of scope of this PR, but we could have some doctest at least in the docstring at least (instead of having a simple comment).

jbertran reviewed (commented):
Some formatting issues, some stray stuff. Please mark comments about stuff remaining to implement as `TODO` so we can clearly differentiate which comments are indicative and which are for ongoing items.

NicolasT commented on salt/_modules/metalk8s_kubernetes.py:
```diff
@@ -2728,3 +2728,29 @@ def show_api_service(name, **kwargs):
         _cleanup(**cfg)
+
+
+def read_storage_class(name, **kwargs):
```
Why would we need this? All required info should be in the pillar.

Ebaneck replied:
@NicolasT I am assuming here that at the start of execution we do not refresh pillar and for some reason, a volume in the pillar has a storage class which no longer exist(volume storageclass was deleted).

`But again, I think pillars get refreshed every now and then`

So,
- We either refresh the pillars and make sure that all volumes in the pillars actually have a valid storage class
- Or Check that each volume to be provisioned has a storage class indeed(reason for the above method).

NicolasT replied:
Volumes in the pillar always have a valid storageclass, right? That's checked in the `ext_pillar`.

The main reason I was asking is as follows: I assume you're adding this code to eventually be used in some state and a state SLS. These SLSs are to be applied on minions, which is in the end where the storage sits.

However.

This code connects to the K8s API. And minions can't connect to the K8s API. Hence all info a minion needs that's derived from info that's stored in the K8s API needs to be passed in as pillar values.

Ebaneck replied:
Thanks for the clarification. It will be useful to simply refresh the pillars hence ensure that all volumes in the pillar are valid(have storageclasses).

NicolasT commented on salt/metalk8s/volume/files/loopdevice.service.j2:
```diff
@@ -0,0 +1,17 @@
+{# source salt linux formulars #}
```
I'm not sure this is the right approach (using a `systemd` unit to set up a loop device).

- We'll have multiple such devices
- How to properly handle cleanup?

Instead, I'd argue we could apply a state at minion startup to get the same thing done.

slaperche-scality replied:
Yes, this may help: https://docs.saltstack.com/en/latest/ref/states/startup.html

Ebaneck replied:
Actually, I see two options here.
- use a startup state as you have suggested or
- use a minion start event 

I am going to experiment with either and see what works well.

NicolasT commented on salt/metalk8s/volume/files/loopdevice.service.j2:
```diff
@@ -0,0 +1,17 @@
+[Service]
+Type=oneshot
+{# The command is prefixed with '-' to consider it a success if the loopback device is already setup #}
+ExecStart=-/sbin/losetup {{ device_name }} {{ file }}
```
How would you figure out a reasonable `device_name`?

Ebaneck replied:
I think we are almost certain that having something like 
`/sbin/losetup/ <volume_uid> <sparsefile_name>` could work no?? Since it is impossible for `device_name` with similar volume_uid or volume_name to exist.

For `device_name` I probably will use the unique `volume_uid`

Ebaneck replied:
Hmmm, seems we will be having multiple loop devices for one particular volume ??

NicolasT commented on salt/metalk8s/volume/prepare/init.sls:
```diff
@@ -0,0 +1,84 @@
+
+Create sparseLoopDevice {{ v_name }} directory:
+  file.directory:
+  - name: /dev/{{ v_name }}
```
Huh, what's this supposed to do?

Ebaneck replied:
Before creating a sparsefile, create a corresponding directory for it in `/dev/volume_name/...`.
`/dev/volume_name/sparsefile_name.img`

NicolasT replied:
You don't ever need to create anything in `/dev` yourself, that's why we have `udev` nowadays. The sparse files are definitely not supposed to be in `/dev`, see other comments.

NicolasT commented on salt/metalk8s/volume/prepare/init.sls:
```diff
@@ -0,0 +1,84 @@
+
+add sparsefile_{{ v_name }}_init_script:
+  file.managed:
+  - name: /usr/lib/systemd/system/loopdevice-{{ v_name }}.service
```
Drop-ins don't go in `/usr`.

Ebaneck replied:
Well noted

I see it should go to `/etc/systemd/system/....`

slaperche-scality commented on salt/metalk8s/volume/prepare/init.sls:
```diff
@@ -0,0 +1,84 @@
+
+Create sparseLoopDevice {{ v_name }} directory:
+  file.directory:
+  - name: /dev/{{ v_name }}
```
Also, sparse files are supposed to live under `/var/lib/metalk8s/storage/sparse/`

Ebaneck replied:
ok

Ebaneck committed:
salt/volume: Installed required dependencies

We need to install e2fsprogs in order to have mkfs.ext4 command.

Refs: #1474

Ebaneck committed:
salt/volume: Add sparseLoopDevice creation

For now, we only handle the preparation of one volume of type
sparseLoopDevice.
In later commit we will handle:
- re-run `losetup` on reboot
- handle rawBlockDevice
- use a module for `mkfs` instead of an ugly shell command
- prepare all volume when none is specified.

Refs: #1474

Ebaneck committed:
salt/volume: add helper to convert k8s quantity to bytes

Volume size are given as `resource.Quantity` objects which are composed
of a number and a unit.
This commit introduce an helper function to convert this quantity into a
number of bytes (that we can pass to the truncate function).

Refs: #1474

slaperche-scality committed:
salt/volume: add module/state for sparse loop device

Simplify the handling of sparse loop device by creating dedicated module
and states.
That way we can keep a simple declarative SLS file, and keep the gritty
details in the Python code.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: fix PrepareVolume

Call the real Salt state to prepare the volume (`metalk8s.volumes`)
instead of calling `test.rand_sleep`.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: add module/state for mkfs

On one hand, the modules provided by Salt don't allow us to specify
arbitrary `mkfs` options.
On the other hand, getting the right command to invoke may be tricky and
require some ugly shell commands/lot of Jinja work.

So let's handle the problem in a dedicated Salt module/state, where we
can use Python code to address it.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: rework volume Salt module

We reworked the internals of the volumes Salt module in order to:
- make the SLS simpler (no more dispatch on the volume type with
  if/else)
- make it easier to support more volume types in the future ("just"
  implement a new class)

To do so, the module now expose a type-agnostic interface of free
functions wich provides the basic required actions (`is_formatted`,
`create`, …).

Then we have an abstract base class that can be inherited for each
Volume type to implement those actions as needed.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: make the volume parameter optional

When no volume names are given, we simply prepare all the volumes
defined on the node.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: add support for rawBlockDevice

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
Current state:
- [x] Prepare a `sparseLoopDevice`
- [x] Prepare a `rawBlockDevice`
- [x] Calling the state with no arguments prepare all the volumes of the node
- [x] Handle reboot for `sparseLoopDevice`

slaperche-scality commented on storage-operator/pkg/salt/client.go:
```diff
@@ -53,27 +53,36 @@ func NewClient(creds *Credential) *Client {
+		"fun":    "state.sls",
+		"kwarg": map[string]interface{}{
+			"mods":    "metalk8s.volumes",
+			"saltenv": "metalk8s-2.4.0-dev",
```
Don't want to hardcode that I guess.

Should I retrieve it from the cluster (with the namespace annotations)?

gdemonet replied:
Indeed, but maybe use the target Node version and not the cluster one? Not sure though :/

TeddyAndrieux replied:
IMO the CRD (and CR, ...) deployed in apiserver should be the one of the cluster_version so I think we should deploy volume using this same version

slaperche-scality replied:
Seen with Teddy: cluster version seems the most pertinent here.

I'll add some code to deal with that.

TeddyAndrieux replied:
Oh sorry misunderstanding what is done in the called state.
Maybe like @gdemonet said we should use the version of the node as it's mainly change on the node.

slaperche-scality replied:
As seen with Nicolas, we want to use the Node version, because that's the one used by the Salt minion so let's stay in sync with it.

NicolasT replied:
This is still outstanding?

slaperche-scality replied:
Yup, but that'll be addressed by my next commit (wasn't ready yet when I made the last push)

slaperche-scality replied:
Done

gdemonet commented on salt/metalk8s/volumes/init.sls:
```diff
@@ -0,0 +1,2 @@
+include:
+  - .prepare
```
Let's try to fix this early: formulas are usually named as describing a **state**, not an **action**. So that would be `prepared.sls`, and could be accompanied by an `unprepared.sls` counterpart.

slaperche-scality replied:
Ok, will rename it to `prepared`.

As for `unprepared` it's is planned (#1475)

Ebaneck committed:
salt/volume: rename prepare to prepared

Formulas are named as state, not as action.

Refs: #1474

Ebaneck committed:
salt/volume: Add a minion startup state

This commit adds a minion startup state to ensure that
all loop devices are always created when a minion reboots.

Note that now we need to always specify a saltenv, thus we update the
bootstrap accordingly.

Closes: #1474

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        """
+        params = self['spec.storageClassName.parameters']
+        # Options are stored as JSON-encoded list.
+        mkfs_options = json.loads(params['mkfsOptions'])
```
What if no `mkfsOptions` are set (which I think should be valid)?

slaperche-scality replied:
Good catch, I'll will use `params.get('mkfsOptions', '[]')` to be safe.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        params = self['spec.storageClassName.parameters']
+        # Options are stored as JSON-encoded list.
+        mkfs_options = json.loads(params['mkfsOptions'])
+        command = ['mkfs']
```
From `man mkfs`:

>        This mkfs frontend is deprecated in favour of filesystem specific mkfs.<type> utils.

We shouldn't use `mkfs`, but only support some 'known' filesystems, and use their respective `mkfs.TYPE` utilities, and the correct options.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        mkfs_options = json.loads(params['mkfsOptions'])
+        command = ['mkfs']
+        command.extend(['-t', params['fsType']])
+        command.extend(['-U', self['metadata.uid']])
```
This may be valid for `mkfs.ext4`, but is not valid for e.g. `mkfs.xfs`, hence comment above.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        command.extend(['-t', params['fsType']])
+        command.extend(['-U', self['metadata.uid']])
+        command.extend(mkfs_options)
+        command.append(self.block_device)
```
In case of `SparseLoopDevice`, we'll want to format the backing file before setting up the loop-device. Or does `udev` get properly triggered after `mkfs` to then create the `/dev/disk/by-uuid/...` symlink?

slaperche-scality replied:
Indeed, need to format before `losetup` otherwise the symlink won't be created by `udev`.
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        command.append(self.block_device)
+        _run_cmd(' '.join(command))
+
+    def __getitem__(self, path):
```
That's a very strange/non-standard behaviour/definition of `__getitem__`, not sure I like it.

slaperche-scality replied:
Yeah, I should probably use a "normal" function for that, instead of overloading the behavior of a well known operator.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+    @property
+    def exists(self):
+        path = self.sparse_file
+        return os.path.isfile(path) and os.path.getsize(path) == self.size
```
Why the size check for an existence check? What happens if the size doesn't match (for whatever reason)?

slaperche-scality replied:
If the size doesn't match that means someone other than us created the file, so I wanted to play it safe and don't touch the file.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+
+    def create(self):
+        # Try to create a sparse file, don't clobber existing one!
+        try:
```
This construct could leak the FD if someone adds code before the next `try` and not taking this into account.

slaperche-scality replied:
Solved by using using a contextmanager.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        # Recent losetup support `--nooverlap` but not the one shipped with
+        # CentOS 7.
+        command = ' '.join(
+            ['losetup', '--find', '--partscan', '--show', self.sparse_file]
```
Why would we need `--partscan`?

slaperche-scality replied:
As seen with you, I misunderstood the role of this option.
We don't need it and I'll remove it (same goes for `--show`).

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+            ['losetup', '--find', '--partscan', '--show', self.sparse_file]
+        )
+        result  = _run_cmd(command)
+        return result['stdout'].strip()
```
What for? I don't think we should ever use the `/dev/loopN` path.

slaperche-scality replied:
Actually the return value is not even used by the caller, so yes let's remove it.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+        return result['stdout'].strip()
+
+    @property
+    def block_device(self):
```
I guess this is used to format. I'd rather not do it that way, but format *what we know* (i.e., what's passed in by a user with a `rawBlockDevice`, or in this case, the backing sparse file, which can be formatted in-place, rather than formatting after creating the block device, which is more failure-prone).

slaperche-scality replied:
Done.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,360 @@
+
+        The volume is formatted according to its StorageClass.
+        """
+        params = self['spec.storageClassName.parameters']
```
I'd argue this needs a failsafe which checks whether the device is formatted already, and if so, errors out.

Otherwise, if someone creates a `rawBlockDevice` `Volume` for `/dev/sda1` by mistake, we could wipe a whole server (root FS at least). Not cool.

slaperche-scality replied:
Indeed & Done.

NicolasT commented on salt/metalk8s/salt/minion/files/minion-99-metalk8s.conf.j2:
```diff
@@ -8,3 +8,10 @@ use_superseded:
+# Prepare volume at startup (required for loop devices persistency).
+startup_states: sls
+sls_list:
+  - /metalk8s/volumes/prepared
```
Is there a way to specify the `saltenv` here instead of setting it globally for the minion?

slaperche-scality replied:
@TeddyAndrieux any idea about that?

TeddyAndrieux replied:
From my last search I not found any way to do this. 

I would prefer also to not define it in the minion config, but after some reflection as it's the version of the minion and when we will upgrade/downgrade this node we will also update the minion configuration.... for me it's ok to have it here. (And btw it's maybe also mandatory for startup_states :stuck_out_tongue: )

TeddyAndrieux replied:
Also with this change if salt handle the mount of the volumes that mean if we do a downgrade to something before 2.4 we will no longer mount our volumes (maybe we don't care, maybe not :nerd_face:)

NicolasT replied:
`kubelet` manages mounts, not Salt.

NicolasT replied:
Meh, it's fine I guess... in the end, the minion should be in that saltenv in the first place anyway.

slaperche-scality replied:
Ok, then let's keep it as is (thanks for the feedback @TeddyAndrieux)

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -0,0 +1,51 @@
+Refresh pillars for all minions:
+  salt.function:
+    - name: saltutil.refresh_pillar
+    - tgt: '*'
```
Why?

slaperche-scality replied:
We wanted to be sure that we had an up-to-date pillar at that point, but I guess it's useless/won't do what we want?

NicolasT replied:
There should be no need. The pillar should be recalculated when `state.sls` is invoked. Also, the `tgt: '*'` is quite unexpected. Not even sure how this can work in a `state.sls`-interpreted SLS.

slaperche-scality replied:
Ok, will remove it then.

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -0,0 +1,51 @@
+  metalk8s_volumes.present:
+    - volume: {{ volume }}
+    - require:
+      - file: Create the sparse file spool
```
Strictly speaking this is only required for `sparseLoopDevice` volumes, but fine.

slaperche-scality replied:
Yeah, it avoids the need for an `if` in the Jinja and it simplifies the looping when you create all the volumes.

We could also update the module to create the parent directories when creating the sparse file, but that's less explicit.

NicolasT commented on salt/metalk8s/volumes/prepared/installed.sls:
```diff
@@ -0,0 +1,9 @@
+include:
+  - metalk8s.repo
+
+Install e2fsprogs:
```
Please also support XFS from the start.

slaperche-scality replied:
Done

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -0,0 +1,51 @@
+    - name: saltutil.refresh_pillar
+    - tgt: '*'
+
+{%- set volumes_to_create = [] %}
```
Instead of this, could we have a `single` SLS which requires a single volume name to be present and process it (which can then be called from the operator through SaltAPI), and have this one loop over everything, and include `single` for every loop iteration with the loop value in the context as `volume`?

I.e. instead of supporting both handling one single volume and all of them in the same SLS/Jinja template, split it up to make things easier to comprehend.

TeddyAndrieux replied:
But you can't pass information between state.
You only can with orchestrate with pillar but I think we don't want to have an orchestrate to prepare volume as this definitly not need any orchestration

NicolasT replied:
Is there no way to 'include' another SLS with a given context dict to render that sub-SLS before pasting in the result?

TeddyAndrieux replied:
Not as far as I know, I think it's good question for saltstack slack :)

TeddyAndrieux replied:
Seems not supported but asked since long time https://github.com/saltstack/salt/issues/8878

slaperche-scality replied:
I guess we will have to keep it like this then, right?

TeddyAndrieux replied:
Yes I think so, except if someone else have an idea

NicolasT replied:
No, except writing code. That's really too bad.

slaperche-scality replied:
So I'll resolve the conversation.

But yes, it's a pity that we can't "reuse" easily a templated SLS into another one :(

NicolasT replied:
(Or using a Jinja macro, but bleh)

NicolasT reviewed (changes requested):
Good stuff, some comments.

slaperche-scality committed:
salt/volume: don't overload __getitem__

Since the implemented behavior differs from the usual one, this may be
surprising to people reading the code.
Let's use an explicit method instead.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: remove useless losetup options

We don't need `--show` because we aren't using the ouptut and we don't
need `--partscan` because we aren't expecting (and don't want to handle)
partitions here.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
We now wandle the reboot for `sparseLoopDevice` so the PR is no longer a draft.

This new patchset started to address some of the comments, but a new patchset is coming soon to address the remaining ones: don't need to re-review now (you can wait the next patchset).

slaperche-scality committed:
salt/volume: add a safety check before formatting

Formatting being a destructive operation which can brick an OS and/or
causes data loss, we want to be really careful here.

This commit adds a check to prevent formatting an already formatted
device, that way if someone mispell `/dev/sdb` as `/dev/sda` they won't
destroy their node :]

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: fix operator permissions

Because I mainly test the operator outside of the cluster (and thus
running as admin:admin) I forgot to update those permissions when I
moved from `test.ping` to real Salt states.

We also need `@jobs` to poll the state of async jobs.

We really need some automated testing here ^^"

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: improve `mkfs` handling

- don't use `mkfs` (it's deprecated), use the filesystem specific utils
  (such as `mkfs.ext4`, `mkfs.xfs`, …) instead
- as a consequence check that the given FS type is supported
- for `sparseLoopDevice`, format the sparse file instead of the
  associated loop device (less racy, we're sure that we are formatting
  what we want to) => replace `block_device` property by `path`.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: fix loop device setup

When creating a sparse loop device, the formatting need to happen
**before** associating the sparse file to a loop device (other `udev`
won't create the symlink under `/dev/disk/by-uuid`).

Since the step is now the last, let's rename `initialization` to
`finalization`.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: don't rely on `disk.fstype`

This command it not reliable: it cannot get the FS type from a sparse
file (fair enough), but in such case it doesn't even fail reliably so
you can't even know that something went wrong…

Calling `salt-call disk.fstype /var/lib/metalk8s/storage/sparse/example`
returns "/dev/sda1      ext4  41152736 7754388  31284864  20% /"…
(because it fallbacks on `df` when `lsblk` fails and that's what `df`
returns…)

Even on an actual block device, it returns garbage if it's not
formatted:
`salt-call disk.fstype /dev/sdb`
returns "devtmpfs       devtmpfs   1932084     0   1932084   0% /dev"

Let's use `blkid` instead…

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

TeddyAndrieux commented on salt/_states/metalk8s_volumes.py:
```diff
@@ -0,0 +1,125 @@
+
+    Args:
+        name   (str): SLS caller name
+        volume (str): Volume name
```
It's not common practice in salt normaly name is something usefull and mandatory for state, it's not only a "name"

For me you should remove the volume argument and use name as volume name

TeddyAndrieux replied:
(And it's the same for other function of this file)

slaperche-scality replied:
Ho thanks for the explanation, yeah I'll get rid of the `volume` parameter and use name instead!

slaperche-scality replied:
Done

TeddyAndrieux commented on salt/_states/metalk8s_volumes.py:
```diff
@@ -0,0 +1,125 @@
+    # Idempotence.
+    if __salt__['metalk8s_volumes.exists'](volume):
+        ret['result'] = True
+        ret['comment'] = 'storage for volume {} already exists'.format(volume)
```
Please capitalize at the beginning of comments

slaperche-scality replied:
My rationale here was that you never now how the log message get mix and matched, and thus having proper sentence (starts with a capital, ends with a period) sometime gives weird message in the end.

But yeah, for Salt that's not really an issue so I'll fix that.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+    @property
+    def path(self):
+        return '/var/lib/metalk8s/storage/sparse/{}'.format(
+            self.get('metadata.name')
```
I'm not entirely sure using `name` is the right approach, I'd argue UUID would be better.

Rationale: one can create a volume with some name, then things can go wrong, one could then (possibly much later) create another volume with the same name, and get a conflict.

slaperche-scality replied:
Yeah UUID is better, the PR started with `name` and I kept it but as you said: UUID is safer;
I'll make the change.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+        open_flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
+        try:
+            with _open_fd(self.path, open_flags) as fd:
+                os.ftruncate(fd, self.size)
```
The calculation of `self.size` should be moved *before* the file is opened/created.

Rationale: if `self.size` calculation fails (e.g., invalid format of the value), the `open` will still have happened, so the file will (zero-length) exist on disk. Now when a user would fix the size field, creation will bail out again (size mismatch).

slaperche-scality replied:
Good catch!! I'll move it before the `open`.

slaperche-scality replied:
Hum, we will still have the issue if the `truncate` fails no?

Instead of moving the calcultation, I'll add a try/except around the `truncate` to `os.unlink` in case of error.

NicolasT replied:
But only if the error is not `EEXIST` or whatever, then... Otherwise it kinda defeats the purpose :D

slaperche-scality replied:
True, but what I had in mind was:

```python
        open_flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
        try:
            with _open_fd(self.path, open_flags) as fd:
                try:
                    os.ftruncate(fd, self.size)
                except Exception:
                    os.unlink(self.path)
                    raise
        except OSError as exn:
            raise Exception('cannot create sparse file at {}: {}'.format(
                self.path, exn
            ))
```
So it should be safe (the unlink only happen if `truncate` (or `self.size`), not for `open`).

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+            ))
+
+    @property
+    def is_finalized(self):
```
`finalized` may not be the right term, actually. To me it relates to being removed (e.g., an object finalizer in a GCd language).

NicolasT replied:
`is_provisioned` and `provision`?

slaperche-scality replied:
Yeah, I also wasn't happy with `finalize`…

I'll use `is_provisioned` and `provision`, way better. Thanks!

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+        result  = _run_cmd(command)
+        return re.search(pattern, result['stdout']) is not None
+
+    def finalize(self):
```
See above.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+        return self.get('spec.rawBlockDevice.devicePath')
+
+    def format(self, force=False):
+        # We format an entire device, not just a partition: we need force=True.
```
That surprises me, to be honest...

slaperche-scality replied:
Me too ! But indeed:

```
# mkfs.ext4 /dev/sdb
mke2fs 1.42.9 (28-Dec-2013)
/dev/sdb is entire device, not just one partition!
Proceed anyway? (y,n)
```

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+    elif 'sparseLoopDevice' in volume['spec']:
+        return SparseLoopDevice(volume)
+    else:
+        raise ValueError('unsupported Volume type')
```
Add the volume name in this message.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+# Cannot use `disk.fstype` from Salt because you can't trust its output…
+#
+# Example:
+# - On a unformated sparse file:
```
Ugh that's terrible :-D

slaperche-scality replied:
yup, very surprised by the behavior there…

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+# So yeah, let's not rely on this…
+def _get_fstype(path):
+    try:
+        res = _run_cmd('blkid {}'.format(path))
```
Do we assume this command is pre-installed or should we force-install it, just like `xfsprogs` and such?

slaperche-scality replied:
Provided by `util-linux`:

```
# rpm -q --whatprovides /usr/sbin/blkid
util-linux-2.23.2-59.el7_6.1.x86_64
```

I guess we can assume it's here because we require it for the pre-flight.

```yaml
kubeadm_preflight:
  mandatory:
    packages:
      - util-linux            # provides nsenter, mount
      - iproute               # provides ip
      - iptables
```

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+def _get_fstype(path):
+    try:
+        res = _run_cmd('blkid {}'.format(path))
+        pat = r'TYPE="(?P<fstype>.+)"'
```
Instead of using `blkid` and a regex to parse its output, consider using `lsblk`:
```
$ lsblk -o FSTYPE -r -n /dev/sda2
ext4
```
for actual block devices. Sadly enough, this seems not to work for files:
```
$ truncate -s 1M volume.dat
$ lsblk -o FSTYPE -r -n ./volume.dat
lsblk: ./volume.dat: not a block device
```

slaperche-scality replied:
The output of `lsblk` seems to be variable as well:
```
# lsblk -o FSTYPE -r -n /dev/sda1
ext4
# lsblk -o FSTYPE -r -n /dev/sda  # Why do I have an empty line in that case? oO

ext4
# lsblk -o FSTYPE -r -n /dev/sdb

```
And yeah, it doesn't work on files (even after the formatting it), so I'll stay on `blkid` I think (since it's already installed anyway), otherwise we can't check for the sparse file.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+        pat = r'TYPE="(?P<fstype>.+)"'
+        match = re.search(pat, res['stdout'])
+        return match.groupdict().get('fstype') if match else None
+    except CommandExecutionError:  # Happens on non-formatted sparse file.
```
Could we then assert the given path is a file?

slaperche-scality replied:
What do you mean?

I'll rewrite this part to be more fine-grained (return code==2 I'll treat as "No FS", return code non-zero and !=2 I'll keep raising).

NicolasT replied:
OK, works for me.

I just wanted to make sure that `CommandExecutionError` (i.e., `rc != 0`) doesn't implicitly mean "no FS present", while there are plenty of other reasons why `blkid` can fail...

NicolasT replied:
Looking at the man page, checking for rc 2 catches quite a lot of things as well, including "impossible to gather any information"...

slaperche-scality replied:
Yeah you're right.

"impossible to gather any information" is a bit like `lsblk` failing on non-block device: in the end you don't know if it's formatted or not :/

I guess we could use `file` but the output format is a mess…

```
# truncate -s 1G bar
# file -sbL bar
data
# truncate -s 2G foo
# mkfs.xfs -f foo > /dev/null
# file -sbL foo
SGI XFS filesystem data (blksz 4096, inosz 512, v2 dirs)
# wipefs -af /dev/sdb > /dev/null
# file -sbL /dev/sdb
data
# mkfs.ext4 -F /dev/sdb > /dev/null
mke2fs 1.42.9 (28-Dec-2013)
# file -sbL /dev/sdb
Linux rev 1.0 ext4 filesystem data, UUID=e3276629-5051-431b-83b5-da808679f77a (extents) (64bit) (large files) (huge files)
```

Could `(?P<fstype>.+) filesystem data` and apply a `tolower` on the result, but that looks more hackish than reliable…

slaperche-scality replied:
Done

NicolasT commented on salt/_states/metalk8s_volumes.py:
```diff
@@ -0,0 +1,125 @@
+    return ret
+
+
+def finalized(name, volume):
```
See above.

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -0,0 +1,430 @@
+        return
+
+    @property
+    def is_formatted(self):
```
A case I believe we're not handling right now, and should do (likely by erroring out): what if the volume exists (and has the expected size and such), it is formatted, but its FS UUID is *not* the UUID we expect it to have?

slaperche-scality replied:
Maybe `is_formatted` should only check the UUID in fact: that's the proof that we have formatted it ourself.

Checking the FS type can be dangerous no?
Imagine we have a StorageClass `Foo` with `fsType` set to `xfs`, we format the device as described.
Later someone change the `fsType` of the `StorageClass` (I think you can't edit `parameters` of a `StorageClass` but then simply delete/recreate a new with the same name…)  to `ext4`.
Now at the next call for this volume (like when you reboot) we will refresh the pillar, store the new version of the StorageClass and see `xfs != ext4` => not formatted.
Then we go to format => "ho it's already formatted so we bail out".
No data loss (thanks to the safeguard we added after the previous reviews) but an ugly failure.

Or I'm missing something and this scenario can't happen?

slaperche-scality replied:
So I think I'll rewrite `is_formatted` to check the UUID only:
- they match => nothing to do
- they mismatch => either we don't have one (so we should go format) or we have another one (we go format but then we detect the existing FS and bail out => we're safe).

slaperche-scality replied:
Done

NicolasT commented on salt/metalk8s/volumes/prepared/init.sls:
```diff
@@ -0,0 +1,47 @@
+    {%- do volumes_to_create.extend(all_volumes.keys()|list) %}
+{%- endif %}
+
+Create the sparse file spool:
```
`spool`? `pool`?

slaperche-scality replied:
I meant `spool` but after checking it seems I had a wrong definition of this word so I'll use `directory` instead (let's keep things simple).

slaperche-scality replied:
Done

slaperche-scality committed:
storage-operator/salt: compute saltenv

Don't hardcode saltenv, instead compute it from the node's MetalK8s
version.
We use the node's version instead of the cluster one because that's also
the one used by the Salt minion.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: use `name` instead of `volume`

Since Salt states have a mandatory name parameter, let's use it to pass
the volume name instead of adding an extra-parameter.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
storage-operator/salt: fix job polling

Until now, we used to rely on the value of `success` to know if a job
was completed successfully or not.
Unfortunately, that's not good enough: you can get `success: true` while
nothing was actually run…
See https://github.com/saltstack/salt/issues/4002

This is a problem because if we think that the `PrepareVolume` step was
a success, we will then try to get the size of the device with
`disk.dump`, but if `PrepareVolume` didn't run then the device doesn't
exists and `disk.dump` will fail => we wrongly move the Volume into a
failed state.

And we easily encounter this issue by creating several volume at
once/within a short time interval.

Try to spawn two state executions at the same time:

    salt-call state.sls metalk8s.volumes saltenv=metalk8s-2.4.0-dev pillar='{"volume": "foo"}'&
    salt-call state.sls metalk8s.volumes saltenv=metalk8s-2.4.0-dev pillar='{"volume": "bar"}'

You'll get:
    local:
        Data failed to compile:
    ----------
        The function "state.sls" is running as PID 32544 and was started at 2019, Aug 14 21:45:18.338129 with jid 20190814214518338129

And when it's done by the operator through the API, here is what we get
for the second job:

    "bootstrap": {
      "return": [
        "The function \"state.sls\" is running as PID 22874 and was started at 2019, Aug 14 20:47:46.622817 with jid 20190814204746622817"
      ],
      "retcode": 1,
      "success": true,
      "out": "highstate"
    }

`success` is true, but `retcode` is non-zero…

So, to be 100% sure that the spawned job succeeded we need to check the
return code.

Why not ignoring `success` and only rely on `retcode`?
Because we want to tell the difference between:
- the job ran but failed (success == false): move the Volume into Failed
  state
- the job failed to run (success == true but retcode != 0), probably
  because one was already running: we can retry later

Note that it seems possible to run states in parallel[1], but our state
may install packages and that seems to be a dangerous things to do in
parallel, so let's avoid…
(But maybe we should move the deps install out of this state, that would
probably make sense…)

[1]: https://docs.saltstack.com/en/latest/ref/states/parallel.html

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: use UUID for `is_formatted`

If we want to be really sure that a device have been formatted by use we
shouldn't check the FS type but the UUID.

In fact, relying on the FS type for `is_formatted` can be wrong.

- a Volume V is created with the StorageClass SC which has `fsType` set
  to `xfs`
- we format the device with XFS
- later, the StorageClass SC is deleted and someone reuse the name to
  create a new StorageClass SC which has `fsType` set to `ext4`
- the server reboot, we reprovision all the volume declared on the node,
  when the pillar is updated it stored the volume V and embed the
  content of the (new) StorageClass SC
- when we check if the device FS type (xfs) match the expected one (from
  the StorageClass, now `ext4`) we get a mismatch and think the device
  is not formatted
- we then call `format`, here the safeguard detect that the device is
  already formated and we bail out and avoid data loss, but we have put
  the Volume V in a Failed state when everything was fine: that's not
  good…

So yeah, we really should really on the UUID instead of the FS type.

Note that this only hold for `is_formatted`, the safeguard in `format`
still need to check if we already have something on the device before
formatting to avoid data loss.

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/volume: use libblkid directly

This commit adds a new Python module to call directly into the blkid
library (using ctypes) instead of relying on the `blkid` command.

Why do we do that?
==================

Because we want to be as sure as possible that when `blkid` returns no
informationg is means "there is no info" and not "maybe there is some
info but I couldn't get them" (because if we mix two cases, we may
actually format something because we thought "ho there is nothing,
davai" whereas the reality is "there was data but I haven't seen them
because reasons…").

Relying on `return code 2 == no data` seems kinda safe, but it's not
100% clear from the man page. Now we could check the source code  (and I
did) but that would only be valid for the current version we look at.

So let's directly attack the library, making sure that any API call will
result in an exception and thus if we end up with "no data" then that
truly means "everything went well, but there was nothing there".

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
Ok, new version ready for review.

This time everything was addressed.

There are two new commits that may require your attention:
1. https://github.com/scality/metalk8s/pull/1513/commits/ba91f008c93112293d2df567a5a62eb7bacf53ae: this one handle the issue of "We can't run two state in parallel"
2. https://github.com/scality/metalk8s/pull/1513/commits/be6d91d88d9ff65653cb33b166eb9ff796661970: as discussed on Slack this one tries to call into `libblkid` directly instead of calling the shell command

About ①, I'm wondering if we could use [parallel states](https://docs.saltstack.com/en/latest/ref/states/parallel.html), but to do so we would need to move the installation of the packages out of this state (where could we put them?).

NicolasT commented on salt/_modules/metalk8s_blkid.py:
```diff
@@ -0,0 +1,140 @@
+import ctypes
+import functools
+
+import metalk8s_blkid_c as blkid_c
```
You can't do this in Salt. It may work, but it's not guaranteed to work.

If you have code that's 'utility code', i.e. not 'real' Salt-module functionality, have it as a 'utils' module and consume it from `__utils__`.

Although in this case I'd merge both files in one.

slaperche-scality replied:
I'll move those into `_utils` then.

I split the code between two files to separate the "low-level" binding that only defines the "interfacing" from the "high-level" binding that provides a Python-like interface on top of it.

But yeah, I can merge those, since I'll already using folds I can use them to keep the separate.

Done

NicolasT commented on salt/_modules/metalk8s_blkid.py:
```diff
@@ -0,0 +1,140 @@
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
```
I'm going to be a bit pedantic, but because this interacts with system-level C stuff, better be safe than sorry.

So, this `probe` thing clearly has a `free_probe` function for cleanup (it's C, after all). However, the current Python interface leaves ways for this release function never to be called:

- I can call `Probe.from_filename` and never use the result in a context manager (i.e., `__exit__` will never be called).
- I can construct a new `Probe` as-is, and never use the result in a context manager.

Since `free_probe` *must* be (promptly) called after `new_probe` is executed, we should change the API to enforce this.

This is, actually, quite simple: instead of making `Probe` a public type, make it 'private' (rename to `_Probe`, only have the `configure` and `probe` members)), then create two top-level functions that enforce use as a context-manager (can use this utility function in `contextlib` with `yield` to implement them, I forgot the name), one `def probe_frome_filename(filepath)` and one `def probe(probe=None)`, and done.

slaperche-scality replied:
Very good point!

I'll rework the interface to enforce correct use.

NicolasT commented on salt/_modules/metalk8s_blkid.py:
```diff
@@ -0,0 +1,140 @@
+    def __exit__(self, type, value, traceback):
+        if self._probe:
+            blkid_c.free_probe(self._probe)
+            self._probe = None
```
This introduces a second 'issue': when an API allows one to create some value, and then use it as a context manager, nothing prohibits reuse of said value after the context-managed exited. However, if a consumer of this API would do so, things will act really weird.

slaperche-scality replied:
Yeah, that's why I added this `need_valid_probe` band-aid thing: to avoid segfault if one reuse the object after the `with`.

But it won't be needed after reworking the API as suggested in another comment.

NicolasT replied:
Making 'invalid state' unrepresentable (through 'expected' use of an API, of course... in Python you can't prohibit someone doing crazy things) always trumps runtime validation :)

NicolasT commented on salt/_modules/metalk8s_blkid.py:
```diff
@@ -0,0 +1,140 @@
+            blkid_c.free_probe(self._probe)
+            self._probe = None
+
+    @_need_valid_probe
```
I guess `need_valid_probe` is related to the face `_probe` *can* be `None`, but never should be.

So, instead of allowing it to be `None`, how about requiring it to be passed to the constructor (but see above, this would be 'private', only used by the context manager functions), in which case it never *can* be `None`, and there's no need to validate this (outside of the constructor), because we tie the lifetime of the probe to the lifetime of this `_Probe` instance (actually, the lifetime of the probe spans the lifetime of the `_Probe` instance).

Granted, this wouldn't prevent nonsense like

```python
p = None
with probe_from_filename("/foo/bar") as q:
    p = q

p.probe()
```

but constructs like the above can't be expected to be OK in general anyway.

slaperche-scality replied:
Removed by reworking the API

NicolasT commented on salt/_modules/metalk8s_blkid.py:
```diff
@@ -0,0 +1,140 @@
+    def configure(self, config):
+        """Configure the probe.
+
+        Valid keys for `config` are:
```
That's a very Javascript-esque calling convention. Why not use plain arguments? Much more explicit.

slaperche-scality replied:
I'll replace the dict by a namedtuple, more explicit (while avoiding long signature/long function call).

NicolasT replied:
That would still require one to look at the documentation instead of the function prototype to know what defaults are.

slaperche-scality replied:
Fair enough. Done

NicolasT commented on salt/_modules/metalk8s_blkid_c.py:
```diff
@@ -0,0 +1,138 @@
+
+soname = ctypes.util.find_library('blkid')
+if soname is None:
+    raise Exception('cannot find blkid library')
```
Would an `ImportError` make sense?

slaperche-scality replied:
Yup, I think it's better to use `ImportError` here.
Done.

NicolasT commented on salt/_modules/metalk8s_blkid_c.py:
```diff
@@ -0,0 +1,138 @@
+
+
+# blkid_probe_enable_superblocks
+probe_enable_superblocks= blkid.blkid_probe_enable_superblocks
```
Missing space before `=`

slaperche-scality replied:
Done

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -13,6 +13,8 @@
 
 from salt.exceptions import CommandExecutionError
 
+import metalk8s_blkid as blkid
```
Again, can't do this. Need to look things up from `__salt__` or something across modules IIRC.

slaperche-scality replied:
Yup, now using `__utils__`.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -180,7 +182,8 @@ def format(self, force=False):
-        if _get_from_blkid(self.path, 'TYPE'):
+        device_info = _get_from_blkid(self.path)
+        if device_info.fstype or device_info.has_partition:
             raise Exception('backing device `{}` already formatted'.format(
```
Exception no longer matches the causes for errors. Split them up in 2 cases with an appropriate error message.

slaperche-scality replied:
Having a FS or a partition table are both a proof that the device is already formatted, no?
Do we really want to make a distinction here?

NicolasT replied:
Yes.

Either it's a device that's already formatted but not 'by us', so there's a 'conflict'.

Or, the device path we got (e.g. in case of a `rawBlockDevice`) is a block device that contains partitions. As such this is more likely a configuration error (someone provided `/dev/sda` whilst it should've been `/dev/sda1`, for example).

Granted, in both cases there's nothing *we* can do except for raising an error. However, the root cause why such exception is raised is quite different, and pointing the user in the right direction ('Already formatted' vs. 'Contains partitions, not something we can format') is useful.

slaperche-scality replied:
Indeed, will do.

NicolasT commented on salt/_modules/metalk8s_volumes.py:
```diff
@@ -382,20 +385,17 @@ def _open_fd(*args, **kwargs):
+        'partitions': True,
+    }
+    with blkid.Probe.from_filename(path) as probe:
+        probe.configure(probe_config)
```
Would it make sense to allow configuration from the 'creation' function directly?

slaperche-scality replied:
Yup, I'll move it there.

Done.

NicolasT commented on salt/_utils/volume_utils.py:
```diff
@@ -0,0 +1,297 @@
+
+    def destroy(self):
+        """Free the resources allocated by the probe."""
+        free_probe(self._probe)
```
And now you should set `self._probe` to None (and check for it before), otherwise one can call `destroy` (which is public API!), and cause a double-free.

NicolasT replied:
I'd argue the constructor should take a `probe` which is created externally to this class code, and then create/destroy the C-level 'probe' inside the context manager code.

slaperche-scality replied:
Done

NicolasT commented on salt/_utils/volume_utils.py:
```diff
@@ -0,0 +1,297 @@
+    )
+
+
+def get_parition_flags(*args):
```
`parition`?

slaperche-scality replied:
meh, typo >_<
Done.

NicolasT commented on salt/metalk8s/salt/master/files/master-99-metalk8s.conf.j2:
```diff
@@ -37,4 +37,7 @@ external_auth:
-        - 'test.ping'
+        - 'test.rand_sleep'
+        - 'disk.dump'
+        - 'state.sls'
```
We can further limit this, please file a ticket.

slaperche-scality replied:
Like that:

```yaml
        - 'test.rand_sleep'
        - 'disk.dump'
        - 'state.sls':
            args:
              - 'metalk8s.volumes'
```
?

NicolasT replied:
Something along those lines. Not required in this PR as far as I'm concerned, but would be nice to limit access.

slaperche-scality replied:
Tracked by #1528

NicolasT commented on storage-operator/deploy/role.yaml:
```diff
@@ -59,3 +60,9 @@ rules:
+  resources:
+  - storageclasses
+  verbs:
+  - '*'
```
Only `GET`, no?

slaperche-scality replied:
I'll try, but yeah that should be enough.

slaperche-scality replied:
Not enough, need `get`, `list` and `watch`.

Done

NicolasT reviewed (changes requested):
Minor, mostly good.

NicolasT commented on storage-operator/pkg/controller/volume/volume_controller.go:
```diff
@@ -213,6 +213,7 @@ func newReconciler(mgr manager.Manager) reconcile.Reconciler {
 		scheme:   mgr.GetScheme(),
 		recorder: mgr.GetRecorder("volume-controller"),
 		salt:     salt.NewClient(getAuthCredential(mgr.GetConfig())),
+		saltenv:  "",
```
This can't be right, similar how `fetchSaltEnv` can't be right accordingly.

This structure is global to the operator. However, we retrieve a `saltenv` from a `Node`, as intended: the `saltenv` to be used when invoking Salt can be different from one `Volume` to another, when those sit on different `Node`s.

So we'd need something like `getNodeSaltenv` or something (not even cached, as far as I'm concerned), and use the resulting value when invoking Salt, within the scope of handling a single `Volume`.

slaperche-scality replied:
It was working fine because `saltenv` was recomputed at each reconciliation, but it was confusing.
I'll remove the field and use a local variable in the reconciliation loop instead.

Done.

NicolasT reviewed (changes requested):
I believe there's one last bug, otherwise LGTM, but would appreciate a second +1.

slaperche-scality committed:
salt/volume: use dedicated state for startup

Retrospectively, using the complete state `metalk8s.volume.prepared`,
while working, isn't a great idea.

Using a dedicated state that only does the "provisioning" as several
benefit:
- it's faster (we don't need to check for depedencies, check if disks
  are formatted, …)
- it's safer (we're sure that we aren't gonna trigger costly formating
  or things like that)
- it can be ran in parallel (thus, it won't block the execution of other
  states while it's running).

Refs: #1474
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on salt/metalk8s/orchestrate/etcd.sls:
```diff
@@ -32,6 +32,7 @@ Check etcd cluster health for {{ node }}:
     - metalk8s_etcd.check_etcd_health:
       - minion_id: {{ node }}
     - attemps: 5
+    - until: "cluster is healthy"
```
Where does this come from?

slaperche-scality replied:
Was from a TEST commit where I was trying stuff to make the expansion working.

The idea was to execute `state.running` to check if something was running after restarting the minion (if not == then startup is over, we can proceed with the other states).

To do so, I had to add the support for `until` to `metalk8s.module_run` and thus update its use here.

But in the end it didn't work, so it was removed ;)

Now trying to use a dedicated parallel state for startup (seems to be working so far).

NicolasT commented on salt/metalk8s/salt/minion/files/minion-99-metalk8s.conf.j2:
```diff
@@ -14,4 +14,4 @@ saltenv: {{ saltenv }}
 startup_states: sls
 sls_list:
-  - /metalk8s/volumes/prepared
+  - /metalk8s/volumes/provisioned
```
Applying this state may easily fail (potentially only partially). Is that handled gracefully by Salt?

NicolasT replied:
What I'm aiming at: it's possible `provision` fails because `exists` is `False` (e.g., we didn't prepare the sparse file yet).

TeddyAndrieux replied:
I think the output of the state is just ignored

NicolasT replied:
If that's the case (and logs aren't being spammed), then that's OK for me.

slaperche-scality replied:
That doesn't prevent the minion to start.

Just have an error logged:

```
[DEBUG   ] minion return: {u'fun_args': [[u'/metalk8s/volumes/provisioned']], u'jid': u'req', u'return': {u'test_|-Provision backing storage for kotlet_|-Provision backing storage for kotlet_|-fail_without_changes': {u'comment': u'Failure!', u'name': u'Provision backing st
orage for kotlet', u'start_time': '15:07:37.884802', u'result': False, u'duration': 0.948, u'__run_num__': 0, u'__sls__': u'/metalk8s/volumes/provisioned', u'changes': {}, u'__id__': u'Provision backing storage for kotlet'}}, u'retcode': 2, u'success': True, u'fun': u'stat
e.sls'}
```

NicolasT commented:
I think this can be +1'd?

slaperche-scality commented:
/approve

bert-e commented:
# Not Author



I'm afraid I cannot do that, @slaperche-scality:

Only the author of the pull request @ebaneck can use this command.

Please **delete** the corresponding comment so I can move on.




*The following options are set:* **wait**

Ebaneck commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
